iree/compiler/Codegen/LLVMGPU/LLVMGPUSwizzle.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+/// This function implements the following swizzling logic
+/// void getTiledId2(unsigned x, unsigned y, unsigned* tiledx,
+///                 unsigned* tiledy) {
+///  unsigned t_tiledx = (x + (y % tile) * grid_size_x) / tile;
+///  unsigned t_tiledy = (y / tile) * tile +
+///      (x + (y % tile) * grid_size_x) % tile;
+///  bool c = grid_size_y % tile != 0 &&
+///      ((y / tile) * tile + tile) > grid_size_y;
+///  *tiledx = c ? x : t_tiledx;
+///  *tiledy = c ? y : t_tiledy;
+/// }
+static void makeSwizzledId(Location loc, OpBuilder b, Value workgroupIdX,
+                           Value workgroupIdY, Value gridSizeX, Value gridSizeY,
+                           Value& SwizzledIdX, Value& SwizzledIdY,
+                           unsigned swizzleTile) {
+  Value zero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value tile = b.create<arith::ConstantIndexOp>(loc, swizzleTile);
+  Value yModTile = b.create<arith::RemUIOp>(loc, workgroupIdY, tile);
```

Comment:
You'd probably want to generate affine ops. That would compose and optimize better in general.

Reply:
affine map handle everything as signed and generates extra ops for division/rem, here we definitely want to keep the number of instructions added to a minimum so I’d prefer leave it like that. (Unless there is a way to generate the same code sequence). 
I don’t think there are really opportunities for combining this arithmetic